rfcs/text/0000-basic-typed-contracts.md

```diff
@@ -0,0 +1,249 @@
+Typed Racket isn't always smooth so we might be able to help the
+Racket programmer if we give a meaningful type to that contract.
+
+Starting with the function `positive?`, its type is `(-> Real
+Boolean)` and it has a positive proposition for `Positive-Real`. So
+based on what we saw in the previous section, it has the contract type
+`(Contract Real Positive-Real)`.
+
+The function contract combinator `->/c` wraps whatever function it is
+applied to, ensuring that a caller provides arguments that satisfy the
+domain contract and that the function returns a result that satisfies
+the range contract; Values flow from the caller through the domain
+contract into the function, and out of the function and through the
+range contract.
+
+This value flow is reflected in the type of a function contract. For
+our `(->/c positive? positive?)` the type is
+
+    (Contract (-> {2} Positive-Real Real {3})
+              (-> {1} Real Positive-Real {4}))
+
+In the output type of the contract, here a function type, the domain
+type is `Real` (marked `{1}`). This is because after applying the
+contract to a function, calling the result with a value will apply the
+domain contract (with input type `Real`) to that value. If the value
+passes the domain contract, we know it is a `Positive-Real` by the
+domain contract's output type, which is reflected in the input type's
+domain contract (marked `{2}`). This means that a function we apply
+this contract to has to accept `Positive-Real` values.
+
+In the input type of the contract, the range of the type is `Real`
+(marked `{3}`) because applying the contract to a function is only
+allowed if that function returns a value that can flow through the
+range contract (whose input type is `Real`). Finally, in the output
+type of the contract the range type is `Positive-Real` (marked `{4}`)
+because that's the output type of the range contract. This means that
+if we apply this contract to a function, then the resuling function
+returns at least a `Positive-Real` if it returns at all.
+
+# Reference-level explanation
+
+## Interaction with other features
+
+As far as I can tell contracts are mostly self-contained. Some
+exceptions:
+
+  - Predicates are subtypes of the `FlatContract` type.
+
+  - Values of `FlatContract` type are not functions (`FlatContract`
+    currently corresponds to Racket's `flat-contract?` which is `#t`
+    for e.g. numbers)
+
+  - Applying a function contract (with the current type rule for
+    contract application) combines the range type of the function
+    contract's output type with the range type of the function that
+    will be monitored. In the code this is called "pairwise
+    intersection" and in the thesis this is the `comb` operator.
+
+As far as subtyping goes, `(Contract S T)` and `(FlatContract S T)` is
+contravariant in `S` and covariant in `T`.
+
+## How the feature would be implemented
+
+The MS thesis linked up top gives type rules formalizing what I walked
```

Comment:
Is there anything to implement that goes beyond what's "standard" for adding a new type?

If not, then I guess this section can say: we need to add the new types, and that means a syntax, printer, typechecking, subtyping, and ~type->contracts~ [oh! guess not]. (Without going into further detail)

Reply:
AFAIK defining the types was pretty standard. RFC updated.